### PR TITLE
Cookie-based authenticated sessions

### PR DESCRIPTION
**Pull request by iamacook**
## Summary

This transitions authorisation from returning an access token that the client need save and handle to using cookie-based sessions meaning that the client only need to include credentials with each request. If verification succeeds, an `access_token` cookie is set for the client that provides authorisation.

In order for this to function, we will need to set the correct origin and allow credentials in the CORS config. As we have varying origins in usage, e.g. local and varying environments, we can dynamically set it to the requester as per [this example](https://stackoverflow.com/questions/51744536/cors-rules-nginx-ingress-rules/61013816#61013816).

Note: this is the same as #1394. I encountered issues in that PR and was unable to recover the diffs after trying to soft reset the commits so this is the replacement.

## Changes

- Install and use `cookie-parser`
- Set access token cookie in `AuthController`
- Replace logic for extracting a token from the `Authorization` header with that from cookies
- Update tests accordingly

### CODE REVIEW

File: src/routes/auth/auth.controller.ts [hunk 1]
```diff
@@ -32,12 +35,26 @@ export class AuthController {
...
+      await this.authService.getAccessToken(verifyAuthMessageDto);
+
+    const maxAge = verifyAuthMessageDto.message.expirationTime
+      ? getMillisecondsUntil(
+          new Date(verifyAuthMessageDto.message.expirationTime),
+        )
+      : undefined;
+
+    res.cookie(AuthController.ACCESS_TOKEN_COOKIE_NAME, accessToken, {
+      httpOnly: true,
```
Comment (fmrsabino) [inline src/routes/auth/auth.controller.ts:53]: One thing that we should consider is the support of HTTP only cookies for non-browser clients (e.g.: mobile). In this case the cookie management won't be handled by the browser but by the HTTP client itself.

From what I know, if we return this in a cookie, this state won't be handled automatically on non-browser clients so additional work needs to be done there.

File: src/routes/auth/auth.controller.ts [hunk 2]
```diff
@@ -32,12 +35,26 @@ export class AuthController {
...
+      secure: true,
+      sameSite: 'lax',
+      path: '/',
+      maxAge,
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/routes/auth/auth.controller.ts:57]: Good point, and I've also realised that the expiration of the token is currently not being set. I will move this back into progress and work on this.

[the remark this answers is not in the document]
Comment (iamacook) [reply src/routes/auth/auth.controller.ts]: The `exp` claim (expiration of the token) is now used as the `MaxAge` as of 02cdac464160e87fa979ce019ebd02ad4f940434 / 7e5b3873b641782dbd999f6be7b2ea0101c8915e.

File: src/routes/auth/auth.controller.ts (hunk 1 above)
Comment (iamacook) [reply src/routes/auth/auth.controller.ts:47]: I'll be sure to make a note of this when documenting the implementation.

File: src/routes/auth/auth.controller.spec.ts [hunk 3]
```diff
@@ -99,13 +97,21 @@ describe('AuthController', () => {
...
         .build();
       const signature = await signer.signMessage({
         message: toSignableSiweMessage(message),
       });
+      const expirationTimeInSeconds = getSecondsUntil(expirationTime);
+      // MaxAge does not include the second of expiration
+      const maxAge = expirationTimeInSeconds - 1;
+      // jsonwebtoken sets expiration based on timespans, not exact dates
+      // meaning we cannot use expirationTime directly
+      const expires = new Date(Date.now() + expirationTimeInSeconds * 1_000);
```
Comment (iamacook) [inline src/routes/auth/auth.controller.spec.ts:114]: These intricacies are simplified in [a follow up PR](https://github.com/safe-global/safe-client-gateway/pull/1414).

**Review by fmrsabino — APPROVED** [review]
👏

File: src/domain/jwt/jwt.repository.ts [hunk 4]
```diff
@@ -5,6 +5,10 @@ import {
   JwtAccessTokenPayload,
   JwtAccessTokenPayloadSchema,
 } from '@/routes/auth/entities/jwt-access-token.payload.entity';
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/domain/jwt/jwt.repository.ts]: Moved in f19f860c853eb2ac2b6fcb70fd29af31968122d6.

File: src/routes/auth/auth.controller.ts [hunk 5]
```diff
@@ -32,12 +35,31 @@ export class AuthController {
...
+  }
+
+  /**
+   * Extract the expiration time from the token and return the maximum age.
+   * @param accessToken - JWT token
+   * @returns maximum age of the token in milliseconds or undefined if none set
+   */
+  private getMaxAge(accessToken: string): number | undefined {
+    const { exp } = this.authService.getTokenPayloadWithClaims(accessToken);
+    return exp ? getMillisecondsUntil(new Date(exp * 1_000)) : undefined;
```
[the remark this answers is not in the document]
Comment (iamacook) [reply src/routes/auth/auth.controller.ts:63]: I don't think so as `Max-Age` is in milliseconds and `exp` is in seconds (according to the JWT standard).

**Review by fmrsabino — APPROVED** [review]
👏 👏 👏

**Pull request merged**